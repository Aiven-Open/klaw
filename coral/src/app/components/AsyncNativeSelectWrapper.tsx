import { NativeSelect, NativeSelectProps, useToast } from "@aivenio/aquarium";
import { isValidElement, ReactElement, ReactNode, useEffect } from "react";
import { parseErrorMsg } from "src/services/mutation-utils";
import { isDevMode } from "src/services/is-dev-mode";

function isNativeSelectComponent(
  child: ReactNode
): child is ReactElement<NativeSelectProps> {
  return (
    isValidElement(child) &&
    /* eslint-disable-next-line  @typescript-eslint/no-explicit-any */
    (child.type as any)?.render?.displayName === NativeSelect.displayName
  );
}

type AsyncNativeSelectWrapperProps = {
  /**
   * `entity` is the entity which user filters by, e.g. "Team" or "Topic"
   * It is used in e.g. placeholder and toast notification for error cases,
   * like "Error loading <ENTITY>"
   */
  entity: string;
  isLoading: boolean;
  isError: boolean;
  error: unknown;
  children: ReactElement<NativeSelectProps>;
};

/** <AsyncNativeSelectWrapper> handles loading
 * and error states for a <NativeSelect> that uses
 * data that is fetched async (e.g. from API)
 * only takes one <NativeSelect /> as a child.
 */
function AsyncNativeSelectWrapper(props: AsyncNativeSelectWrapperProps) {
  const { entity, isLoading, isError, error, children } = props;

  const toast = useToast();

  // Type-guard to make sure we're only passing
  // <NativeSelect> as a child component
  useEffect(() => {
    if (!isNativeSelectComponent(children)) {
      const errorMessage =
        "Invalid child component. `AsyncNativeSelectWrapper` only accepts `NativeSelect` as a child.";

      if (isDevMode()) {
        throw new Error(errorMessage);
      } else {
        console.error(errorMessage);
      }
    }
  }, [children]);

  useEffect(() => {
    if (isError) {
      toast({
        message: `Error loading ${entity}: ${parseErrorMsg(error)}`,
        position: "bottom-left",
        variant: "default",
      });
    }
  }, [isError]);

  if (isError) {
    return (
      <NativeSelect
        valid={false}
        disabled={true}
        labelText={children.props.labelText}
        aria-label={`No ${entity}`}
        placeholder={`No ${entity}`}
        helperText={`${entity} could not be loaded.`}
      />
    );
  }

  if (isLoading) {
    return (
      <div data-testid={"async-select-loading"}>
        <NativeSelect.Skeleton />
      </div>
    );
  }

  return children;
}

export { AsyncNativeSelectWrapper };
