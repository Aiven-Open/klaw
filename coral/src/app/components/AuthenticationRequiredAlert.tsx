import { Box, Typography } from "@aivenio/aquarium";

function AuthenticationRequiredAlert() {
  return (
    <Box
      role="alertdialog"
      aria-labelledby={"authentication-required-heading"}
      aria-describedby={"authentication-required-text"}
      display={"flex"}
      flexDirection={"column"}
      paddingTop={"l6"}
      justifyContent={"center"}
      alignItems={"center"}
    >
      <Typography.Heading color={"secondary-100"}>
        <span id={"authentication-required-heading"}>
          Authentication session expired
        </span>
      </Typography.Heading>

      <Typography.LargeText>
        <span id={"authentication-required-text"}>
          Redirecting to login page.
        </span>
      </Typography.LargeText>
    </Box>
  );
}

export { AuthenticationRequiredAlert };
