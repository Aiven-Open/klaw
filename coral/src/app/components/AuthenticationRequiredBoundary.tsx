import { Box, Typography } from "@aivenio/aquarium";
import { Component, ReactElement } from "react";
import { isUnauthorizedError } from "src/services/api";

class AuthenticationRequiredBoundary extends Component<{
  children: ReactElement;
}> {
  public state = {
    hasError: false,
  };
  constructor(props: { children: ReactElement }) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error: unknown) {
    if (isUnauthorizedError(error)) {
      return { hasError: true };
    }
  }

  render() {
    if (this.state.hasError) {
      window.location.assign("/login");
      return (
        <Box
          role="alertdialog"
          aria-labelledby={"authentication-required-heading"}
          aria-describedby={"authentication-required-text"}
          display={"flex"}
          flexDirection={"column"}
          paddingTop={"l6"}
          justifyContent={"center"}
          alignItems={"center"}
        >
          <Typography.Heading color={"secondary-100"}>
            <span id={"authentication-required-heading"}>
              Authentication session expired
            </span>
          </Typography.Heading>

          <Typography.LargeText>
            <span id={"authentication-required-text"}>
              Redirecting to login page.
            </span>
          </Typography.LargeText>
        </Box>
      );
    }

    return this.props.children;
  }
}

export default AuthenticationRequiredBoundary;
