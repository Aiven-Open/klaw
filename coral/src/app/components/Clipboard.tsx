import { Tooltip, PositionerPlacement, Button } from "@aivenio/aquarium";
import duplicate from "@aivenio/aquarium/icons/duplicate";
import { useRef, useState, useEffect, FormEvent } from "react";

const copyToClipboard = async (text: string): Promise<void> => {
  try {
    await navigator.clipboard.writeText(text);
  } catch (error) {
    console.error("Failed to copy to clipboard:", error);
  }
};

const ClipBoard = ({
  text,
  accessibleCopyDescription,
  accessibleCopiedDescription,
  description,
}: {
  text: string;
  accessibleCopyDescription: string;
  accessibleCopiedDescription: string;
  description?: string;
}) => {
  const feedbackTimerRef = useRef<number>();

  const [showCopyFeedback, setShowCopyFeedback] = useState(false);

  const clearTimeouts = () => {
    window.clearTimeout(feedbackTimerRef.current);
  };

  useEffect(() => () => clearTimeouts(), []);

  function handleCopy(event: FormEvent) {
    event.preventDefault();
    copyToClipboard(text);
    setShowCopyFeedback(true);
    feedbackTimerRef.current = window.setTimeout(
      () => setShowCopyFeedback(false),
      1000
    );
  }

  if (showCopyFeedback) {
    return (
      <Tooltip
        key="copied"
        content="Copied"
        isOpen
        placement={PositionerPlacement.top}
      >
        <Button.SecondaryGhost
          key="copy-button"
          aria-label={accessibleCopiedDescription}
          onClick={handleCopy}
          icon={duplicate}
        >
          {description}
        </Button.SecondaryGhost>
        <div aria-live="polite" className={"visually-hidden"}>
          {accessibleCopiedDescription}
        </div>
      </Tooltip>
    );
  }

  return (
    <Tooltip
      key="copy-to-clipboard"
      content="Copy"
      placement={PositionerPlacement.top}
    >
      <Button.SecondaryGhost
        key="copy-button"
        aria-label={accessibleCopyDescription}
        onClick={handleCopy}
        icon={duplicate}
      >
        {description}
      </Button.SecondaryGhost>
    </Tooltip>
  );
};

export default ClipBoard;
