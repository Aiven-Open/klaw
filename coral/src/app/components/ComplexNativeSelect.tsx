import { NativeSelect } from "@aivenio/aquarium";
import { ChangeEvent } from "react";

type ComplexNativeSelectProps<T> = {
  options: Array<T>;
  onBlur: (option: T | undefined) => void;
  optionToString: (opt: T) => string;
  getValue: (opt: T) => string;
  placeholder: string;
  labelText: string;
  disabled?: boolean;
  defaultValue?: string;
  required?: boolean;
  error?: string;
};

function ComplexNativeSelect<T>(props: ComplexNativeSelectProps<T>) {
  const {
    options,
    onBlur,
    disabled = false,
    required = false,
    defaultValue,
    labelText,
    optionToString,
    placeholder,
    getValue,
    error,
  } = props;

  // the placeholder behavior will be covered by DS NativeSelect
  // soon, this is a temp solution
  const placeholderValue = "055d87f2-8cd4-11ed-a1eb-0242ac120002";
  function setNewOption({ target: { value } }: ChangeEvent<HTMLSelectElement>) {
    if (value === placeholderValue) {
      onBlur(undefined);
    } else {
      const newOption = options.find((option) => getValue(option) === value);
      if (newOption) {
        onBlur(newOption);
      }
    }
  }

  return (
    <NativeSelect
      onBlur={setNewOption}
      disabled={disabled}
      labelText={labelText}
      required={required}
      defaultValue={defaultValue || placeholderValue}
      valid={error ? false : undefined}
      helperText={error}
    >
      {!defaultValue && (
        <option key={placeholderValue} value={placeholderValue} disabled={true}>
          {placeholder}
        </option>
      )}
      {options.map((option) => {
        const value = getValue(option);
        return (
          <option key={value} value={value}>
            {optionToString(option)}
          </option>
        );
      })}
    </NativeSelect>
  );
}

export { ComplexNativeSelect };
export type { ComplexNativeSelectProps };
