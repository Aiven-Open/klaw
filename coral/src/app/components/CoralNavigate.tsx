import { Navigate, NavigateProps } from "react-router-dom";

type Props = Omit<NavigateProps, "to"> & {
  to: `/${string}`;
  useLegacy: boolean;
};

/**
 * CoralNavigate is a convenience component, which provides unified API
 * to do navigation either with React Router or by using window.location API
 *
 * Once the Reactification is complete, we should migrate all instances of
 * <CoralNavigate /> to use the React Router <Navigate />
 */
function CoralNavigate({ useLegacy, to, ...navigateProps }: Props) {
  if (useLegacy) {
    window.location.assign(to);
    return <></>;
  }
  return <Navigate to={to} {...navigateProps} />;
}

export default CoralNavigate;
