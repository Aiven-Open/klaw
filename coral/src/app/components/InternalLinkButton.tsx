import { asButton, ButtonProps, Button } from "@aivenio/aquarium";
import { Link, LinkProps } from "react-router-dom";
import { ResolveIntersectionTypes } from "types/utils";
import { ReactNode } from "react";

const AsButton = asButton<"a", LinkProps, HTMLAnchorElement>(Link);

type InternalLinkButtonProps = ResolveIntersectionTypes<
  {
    children: string | ReactNode;
    kind?: ButtonProps["kind"];
    disabled?: boolean;
  } & LinkProps
>;
function InternalLinkButton({ children, ...props }: InternalLinkButtonProps) {
  if (props.disabled) {
    // a Link cannot really be disabled semantically. Adding an empty ref
    // with aria-disabled would solve this, but that there are no styles
    // from DS to cover this, so we're showing a disabled button with role
    // link in this case instead to provide the best feedback for assistive technology
    return (
      <Button
        role={"link"}
        aria-disabled={true}
        disabled={true}
        kind={props.kind}
      >
        {children}
      </Button>
    );
  }
  return <AsButton {...props}>{children}</AsButton>;
}

export { InternalLinkButton };
