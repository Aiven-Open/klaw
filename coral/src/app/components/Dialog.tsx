import { Modal, ModalProps } from "src/app/components/Modal";
import { DialogType } from "@aivenio/aquarium/atoms/Dialog/Dialog";
import confirm from "@aivenio/aquarium/dist/src/icons/confirm";
import warningSign from "@aivenio/aquarium/dist/src/icons/warningSign";
import error from "@aivenio/aquarium/dist/src/icons/error";
import data from "@aivenio/aquarium/icons/list";
import { Box, Icon, Typography } from "@aivenio/aquarium";
import { ResolveIntersectionTypes } from "types/utils";

type DialogProps = ResolveIntersectionTypes<
  Omit<ModalProps, "isDialog" | "dialogTitle" | "close"> &
    Required<Pick<ModalProps, "secondaryAction">> & {
      type: DialogType;
    }
>;

const dialogTypeMap: Record<
  DialogType,
  { icon: typeof data; color: "info-70" | "secondary-70" | "error-70" }
> = {
  confirmation: {
    icon: confirm,
    color: "info-70",
  },
  warning: {
    icon: warningSign,
    color: "secondary-70",
  },
  danger: {
    icon: error,
    color: "error-70",
  },
};

function Dialog(props: DialogProps) {
  const { type, title } = props;

  const dialogTitle = (
    <Typography.LargeStrong
      htmlTag={"h1"}
      id={"modal-focus"}
      data-focusable
      color={dialogTypeMap[type].color}
    >
      <Box
        component={"span"}
        display={"flex"}
        alignItems={"center"}
        colGap={"l1"}
      >
        <Icon icon={dialogTypeMap[type].icon} />
        <span>{title}</span>
      </Box>
    </Typography.LargeStrong>
  );

  return (
    <Modal
      title={""}
      dialogTitle={dialogTitle}
      isDialog={true}
      primaryAction={props.primaryAction}
      secondaryAction={props.secondaryAction}
    >
      <>{props.children}</>
    </Modal>
  );
}

export { Dialog };
