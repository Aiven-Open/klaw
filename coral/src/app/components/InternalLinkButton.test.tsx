import { cleanup, screen } from "@testing-library/react";
import { InternalLinkButton } from "src/app/components/InternalLinkButton";
import { customRender } from "src/services/test-utils/render-with-wrappers";
import { mockIntersectionObserver } from "src/services/test-utils/mock-intersection-observer";

describe("InternalLinkButton", () => {
  mockIntersectionObserver();
  const testTo = "/test/url";
  const testText = "My nice link";

  describe("renders all necessary elements", () => {
    beforeAll(() => {
      customRender(
        <InternalLinkButton to={testTo}>{testText}</InternalLinkButton>,
        { browserRouter: true }
      );
    });

    afterAll(cleanup);

    it("shows a link", () => {
      const link = screen.getByRole("link", { name: testText });

      expect(link).toBeVisible();
    });

    it("renders a given route as 'href'", () => {
      const link = screen.getByRole("link", { name: testText });

      expect(link).toHaveAttribute("href", testTo);
    });
  });

  describe("renders different styles based on selected Aquarium's Button props", () => {
    afterEach(cleanup);

    it("shows a primary button by default", () => {
      const { container } = customRender(
        <InternalLinkButton to={testTo}>{testText}</InternalLinkButton>,
        { browserRouter: true }
      );

      expect(container).toMatchSnapshot();
    });

    it("shows a secondary button based on props", () => {
      const { container } = customRender(
        <InternalLinkButton to={testTo} kind={"secondary"}>
          {testText}
        </InternalLinkButton>,
        { browserRouter: true }
      );

      expect(container).toMatchSnapshot();
    });
  });
});
