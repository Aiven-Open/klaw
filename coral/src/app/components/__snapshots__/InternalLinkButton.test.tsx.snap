// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InternalLinkButton renders different styles based on selected Aquarium's Button props shows a primary button by default 1`] = `
<div>
  <a
    class="Aquarium-Button Aquarium-Button.Primary text-white bg-primary-80 active:bg-primary-90 active:ring-0 focus-visible:ring-2 focus-visible:ring-primary-100 focus-visible:ring-inset hover:bg-primary-90 disabled:bg-primary-40 inline-block border-0 rounded-sm transition whitespace-nowrap focus:outline-none relative text-center typography-default-strong py-3 px-4 icon-stroke-2"
    href="/test/url"
  >
    My nice link
  </a>
</div>
`;

exports[`InternalLinkButton renders different styles based on selected Aquarium's Button props shows a secondary button based on props 1`] = `
<div>
  <a
    class="Aquarium-Button Aquarium-Button.Secondary text-primary-80 bg-white ring-1 ring-primary-80 ring-inset active:bg-primary-5 active:ring-primary-90 active:text-primary-90 focus-visible:ring-2 focus-visible:ring-primary-90 focus-visible:text-primary-80 hover:bg-primary-5 hover:ring-primary-90 hover:text-primary-90 disabled:bg-white disabled:text-primary-40 disabled:ring-primary-40 inline-block border-0 rounded-sm transition whitespace-nowrap focus:outline-none relative text-center typography-default-strong py-3 px-4 icon-stroke-2"
    href="/test/url"
  >
    My nice link
  </a>
</div>
`;
